File: q26-50/q42.py
# 2020.08.12
# Problem Statement:
# https://leetcode.com/problems/trapping-rain-water/

class Solution:
    def trap(self, height: List[int]) -> int:
        # idea:
        # only if exists higher num for both left and right
        # that index has water above
        # where the area of that index = min(highest in the left, highest in the right) - height[index]
        
        # check 0 answer cases, do early return
        if len(height) <= 2: return 0
        
        # initialize answer to return
        area = 0
        
        # initialize the left-highest and right-highest for index 1
        left = height[0]
        right = max(height[2: ])
        
        for i in range(1, len(height)-1):
            # update left and right
            if height[i-1] > left:
                left = height[i-1]
            if height[i] == right and height[i] not in height[i+1: ]:
                right = max(height[i+1: ])
            # calculate area for that index
            if left > height[i] and right > height[i]:
                area = area + min(left, right) - height[i]
                
        return area

File: Amazon/q42.java
// 2022.05.17
// Problem Statement:
// https://leetcode.com/problems/trapping-rain-water/

// idea: for each position, take zero rain if larger than either left or right max,
// otherwise, take min(left max, right max) - itself rain
class Solution {
    public int trap(int[] height) {
        if (height.length<=2) return 0;
        int [] left_max = new int [height.length];
        int [] right_max = new int [height.length];
        int l_m = 0;
        int r_m = 0;
        left_max[0] = 0; // doesn't matter
        right_max[height.length-1] = 0; // doesn't matter
        for (int i=1; i<height.length; i++) {
            l_m = Math.max(l_m, height[i-1]);
            left_max[i] = l_m;
            r_m = Math.max(r_m, height[height.length-i]);
            right_max[height.length-1-i] = r_m;
        }
        
        int answer = 0;
        for (int i=1; i<height.length-1; i++) {
            if (height[i]<left_max[i] && height[i]<right_max[i]) {
                answer += Math.min(left_max[i], right_max[i])-height[i];
            }
        }
        return answer;
    }
}